sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,152 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"path"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/inotify"
+)
+
+type inotifyWatcher struct {
+	wd        *inotify.Watcher
+	addWatch  chan *groupToWatch
+	groupList []*groupToWatch
+	pathList  map[string]int32
+}
+
+type groupToWatch struct {
+	name    string
+	folders []string
+	channel chan string
+}
+
+var currentWatcher *inotifyWatcher = &inotifyWatcher{
+	wd:       nil,
+	pathList: make(map[string]int32),
+	addWatch: make(chan *groupToWatch),
+}
+
+func addWatch(newWatch *groupToWatch) {
+	var folderList []string
+	for _, fullPath := range newWatch.folders {
+		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
+		// isn't triggered, so we must monitor the parent folder and use InDelete
+		basePath := path.Dir(fullPath)
+		if _, exists := currentWatcher.pathList[basePath]; !exists {
+			currentWatcher.pathList[basePath] = 0
+			if err := currentWatcher.wd.AddWatch(basePath, inotify.InDelete); err != nil {
+				delete(currentWatcher.pathList, basePath)
+				continue
+			}
+		}
+		currentWatcher.pathList[fullPath]++
+		if osutil.FileExists(fullPath) {
+			folderList = append(folderList, fullPath)
+		} else {
+			removePath(fullPath)
+		}
+	}
+	if len(folderList) == 0 {
+		newWatch.channel <- newWatch.name
+	} else {
+		newWatch.folders = folderList
+		currentWatcher.groupList = append(currentWatcher.groupList, newWatch)
+	}
+}
+
+func removePath(fullPath string) {
+	currentWatcher.pathList[fullPath]--
+	if currentWatcher.pathList[fullPath] == 0 {
+		currentWatcher.wd.RemoveWatch(fullPath)
+		delete(currentWatcher.pathList, fullPath)
+	}
+}
+
+func processEvent(watch *groupToWatch, event *inotify.Event) bool {
+	var tmpFolders []string
+	for _, fullPath := range watch.folders {
+		if fullPath != event.Name {
+			tmpFolders = append(tmpFolders, fullPath)
+		} else {
+			removePath(fullPath)
+		}
+	}
+	watch.folders = tmpFolders
+	if len(tmpFolders) == 0 {
+		watch.channel <- watch.name
+		return false
+	}
+	return true
+}
+
+func watcherMainLoop() {
+	for {
+		select {
+		case event := <-currentWatcher.wd.Event:
+			if event.Mask&inotify.InDelete == 0 {
+				continue
+			}
+			var newGroupList []*groupToWatch
+			for _, watch := range currentWatcher.groupList {
+				if processEvent(watch, event) {
+					newGroupList = append(newGroupList, watch)
+				}
+			}
+			currentWatcher.groupList = newGroupList
+		case newWatch := <-currentWatcher.addWatch:
+			addWatch(newWatch)
+		}
+	}
+}
+
+// MonitorFullDelete allows to monitor a group of files/folders
+// and, when all of them have been deleted, emits the specified name through the channel.
+func monitorFullDelete(name string, folders []string, channel chan string) error {
+	if currentWatcher.wd == nil {
+		wd, err := inotify.NewWatcher()
+		if err != nil {
+			return err
+		}
+		currentWatcher.wd = wd
+		go watcherMainLoop()
+	}
```

Comment:
This is racy if we've got two goroutines that call the function at once and the watcher hasn't been set up. A [`sync.Once`](https://pkg.go.dev/sync#Once) should resolve the race.

Reply:
Good catch!!! Fixed.